client/client.go

```diff
@@ -22,57 +22,318 @@ limitations under the License.
 package client
 
 import (
+	"errors"
 	"fmt"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"sync"
 
 	"github.com/fsouza/go-dockerclient"
+	"github.com/intelsdi-x/snap-plugin-collector-docker/config"
+	"github.com/intelsdi-x/snap-plugin-collector-docker/fs"
+	"github.com/intelsdi-x/snap-plugin-collector-docker/network"
+	"github.com/intelsdi-x/snap-plugin-collector-docker/wrapper"
 	"github.com/opencontainers/runc/libcontainer/cgroups"
 )
 
-const endpoint string = "unix:///var/run/docker.sock"
+const (
+	endpoint         string = "unix:///var/run/docker.sock"
+	dockerVersionKey string = "Version"
+)
 
-type ContainerInfo struct {
-	Id      string
-	Status  string
-	Created int64
-	Image   string
+type DockerClientInterface interface {
+	ListContainersAsMap() (map[string]docker.APIContainers, error)
+	GetStatsFromContainer(string, bool) (*wrapper.Statistics, error)
+	InspectContainer(string) (*docker.Container, error)
+	FindCgroupMountpoint(string) (string, error)
+	Version() ([]int, error)
 }
 
-type DockerClientInterface interface {
-	ListContainers() ([]ContainerInfo, error)
-	FindCgroupMountpoint(subsystem string) (string, error)
+type dockerClient struct {
+	cl           *docker.Client
+	inspectCache map[string]*docker.Container
+	inspectMutex sync.Mutex
 }
 
-type dockerClient struct{}
+type deviceInfo struct {
+	device string
+	major  string
+	minor  string
+}
+
+func init() {
+	getDockerVersion := func() []int {
+		dc := NewDockerClient()
+		version, err := dc.Version()
+		if err != nil {
+			panic(err)
+		}
+		return version
+	}
+	config.DockerVersion = getDockerVersion()
+	fmt.Fprintf(os.Stderr, "Docker runtime version: %v \n", config.DockerVersion)
+}
 
 func NewDockerClient() *dockerClient {
-	return new(dockerClient)
+	client, err := docker.NewClient(endpoint)
+	if err != nil {
+		panic(err)
+	}
+
+	return &dockerClient{cl: client,
+		inspectCache: map[string]*docker.Container{},
+	}
 }
 
 func (dc *dockerClient) FindCgroupMountpoint(subsystem string) (string, error) {
 	return cgroups.FindCgroupMountpoint(subsystem)
 }
 
-func (dc *dockerClient) ListContainers() ([]ContainerInfo, error) {
-	cl, err := docker.NewClient(endpoint)
+func (dc *dockerClient) InspectContainer(id string) (*docker.Container, error) {
+	dc.inspectMutex.Lock()
+	defer dc.inspectMutex.Unlock()
+	if info, haveInfo := dc.inspectCache[id]; !haveInfo {
+		if info, haveInfo = dc.inspectCache[id]; haveInfo {
+			// double-locking is fine here, as side effects are harmless
+			return info, nil
+		}
+		var err error
+		if info, err = dc.cl.InspectContainer(id); err != nil {
+			return nil, err
+		} else {
```

Comment:
`else` block seems to be unnecessary because the previous `if` block ends with return statement